pkg/piecestore/rpc/server/retrieve.go

```diff
@@ -81,59 +85,90 @@ func (s *Server) retrieveData(ctx context.Context, stream pb.PieceStoreRoutes_Re
-	defer func() {
-		baWriteErr := s.DB.WriteBandwidthAllocToDB(latestBA)
-		if latestBA != nil && baWriteErr != nil {
-			log.Println("WriteBandwidthAllocToDB Error:", baWriteErr)
-		}
-	}()
+	allocationTracking := sync2.NewThrottle()
+	totalAllocated := int64(0)
+
+	// Bandwidth Allocation recv loop
+	go func() {
+		var lastTotal int64
+		var lastAllocation *pb.RenterBandwidthAllocation
+		defer func() {
+			if lastAllocation == nil {
+				return
+			}
+			err := s.DB.WriteBandwidthAllocToDB(lastAllocation)
+			if err != nil {
+				// TODO: handle error properly
+				log.Println("WriteBandwidthAllocToDB Error:", err)
+			}
+		}()
 
-	for am.Used < length {
-		// Receive Bandwidth allocation
-		recv, err := stream.Recv()
-		if err != nil {
-			return am.Used, am.TotalAllocated, err
-		}
+		for {
+			recv, err := stream.Recv()
+			if err != nil {
+				allocationTracking.Fail(err)
+				return
+			}
 
-		ba := recv.GetBandwidthallocation()
+			alloc := recv.GetBandwidthallocation()
+			allocData := &pb.RenterBandwidthAllocation_Data{}
+			if err = proto.Unmarshal(alloc.GetData(), allocData); err != nil {
+				allocationTracking.Fail(err)
+				return
+			}
 
-		baData := &pb.RenterBandwidthAllocation_Data{}
+			if err = s.verifySignature(ctx, alloc); err != nil {
+				allocationTracking.Fail(err)
+				return
+			}
 
-		if err = proto.Unmarshal(ba.GetData(), baData); err != nil {
-			return am.Used, am.TotalAllocated, err
-		}
+			// TODO: break when lastTotal >= allocData.GetPayer_allocation().GetData().GetMax_size()
 
-		if err = s.verifySignature(ctx, ba); err != nil {
-			return am.Used, am.TotalAllocated, err
-		}
+			if lastTotal > allocData.GetTotal() {
+				allocationTracking.Fail(fmt.Errorf("got lower allocation was %v got %v", lastTotal, allocData.GetTotal()))
+				return
+			}
 
-		am.NewTotal(baData.GetTotal())
+			atomic.StoreInt64(&totalAllocated, allocData.GetTotal())
 
-		// The bandwidthallocation with the higher total is what I want to earn the most money
-		if baData.GetTotal() > latestTotal {
-			latestBA = ba
-			latestTotal = baData.GetTotal()
+			if err = allocationTracking.Produce(allocData.GetTotal() - lastTotal); err != nil {
+				return
+			}
+
+			lastAllocation = alloc
+			lastTotal = allocData.GetTotal()
 		}
+	}()
 
-		sizeToRead := am.NextReadSize()
+	// Data send loop
+	messageSize := int64(32 << 10)
+	used := int64(0)
 
-		// Write the buffer to the stream we opened earlier
-		n, err := io.CopyN(writer, storeFile, sizeToRead)
-		if n > 0 {
-			if allocErr := am.UseAllocation(n); allocErr != nil {
-				log.Println(allocErr)
-			}
+	for used < length {
+		nextMessageSize, err := allocationTracking.ConsumeOrWait(messageSize)
+		if err != nil {
+			allocationTracking.Fail(err)
+			break
 		}
 
+		used += nextMessageSize
+		n, err := io.CopyN(writer, storeFile, nextMessageSize)
+		// correct errors when needed
+		if n != nextMessageSize {
+			if err := allocationTracking.Produce(nextMessageSize - n); err != nil {
```

Comment:
i'm nervous about future refactors here stomping on the 'err' variable with a scope change. maybe can we make the variable be different than err on this one line?